packages/jupyter-widgets/src/manager/widget-loader.ts

```diff
@@ -0,0 +1,107 @@
+/**
+ * Several functions in this file are based off the html-manager in jupyter-widgets project -
+ * https://github.com/jupyter-widgets/ipywidgets/blob/master/packages/html-manager/src/libembed-amd.ts
+ */
+
+import * as base from "@jupyter-widgets/base";
+import * as controls from "@jupyter-widgets/controls";
+
+let cdn = "https://unpkg.com";
+
+/**
+ * Constructs a well formed module URL for requireJS 
+ * mapping the modulename and version from the base CDN URL
+ * @param moduleName Name of the module corresponding to the widget package
+ * @param moduleVersion Module version returned from kernel
+ */
+function moduleNameToCDNUrl(moduleName: string, moduleVersion: string): string {
+  let packageName = moduleName;
+  let fileName = "index.js"; // default filename
+  // if a '/' is present, like 'foo/bar', packageName is changed to 'foo', and path to 'bar'
+  // We first find the first '/'
+  let index = moduleName.indexOf('/');
+  if (index !== -1 && moduleName[0] === '@') {
+    // if we have a namespace, it's a different story
+    // @foo/bar/baz should translate to @foo/bar and baz
+    // so we find the 2nd '/'
+    index = moduleName.indexOf('/', index + 1);
+  }
+  if (index !== -1) {
+    fileName = moduleName.substr(index + 1);
+    packageName = moduleName.substr(0, index);
+  }
+  const moduleNameString = moduleVersion ? `${packageName}@${moduleVersion}` : packageName;
+  return `${cdn}/${moduleNameString}/dist/${fileName}`;
+}
+
+/**
+ * Load a package using requirejs and return a promise
+ *
+ * @param pkg Package name or names to load
+ */
+function requirePromise(pkg: string | string[]): Promise<any> {
+  return new Promise((resolve, reject) => {
+    const require = (window as any).requirejs;
+    if (require === undefined) {
+      reject('Requirejs is needed, please ensure it is loaded on the page.');
+    }
+    else {
+      // tslint:disable-next-line: non-literal-require
+      require(pkg, resolve, reject);
+    }
+  });
+};
+
+/**
+ * Initialize dependencies that need to be preconfigured for requireJS module loading
+ * Here, we define the jupyter-base, controls package that most 3rd party widgets depend on
+ */
+export function initRequireDeps(){
```

Comment:
What is this for? Seems like it solves the error I was running into in https://github.com/nteract/outputs/pull/11?

Reply:
That's correct. By default, when we attempt to `require` a module, the library would try to fetch the dependencies for that module as well through the same CDN endpoint. Since, we already bundle the base packages, we can just `define` them in the require scope.